protos/icmp/icmp.go

```diff
@@ -0,0 +1,329 @@
+		return IcmpDirectionFromInside
+	}
+	return IcmpDirectionLocalOnly
+}
+
+func (icmp *Icmp) isLocalIp(ip net.IP) bool {
+	if ip.IsLoopback() {
+		return true
+	}
+
+	for _, localIp := range icmp.localIps {
+		if ip.Equal(localIp) {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (icmp *Icmp) getTransaction(k HashableIcmpTuple) *IcmpTransaction {
+	v := icmp.transactions.Get(k)
+	if v != nil {
+		return v.(*IcmpTransaction)
+	}
+	return nil
+}
+
+func (icmp *Icmp) deleteTransaction(k HashableIcmpTuple) *IcmpTransaction {
+	v := icmp.transactions.Delete(k)
+	if v != nil {
+		return v.(*IcmpTransaction)
+	}
+	return nil
+}
+
+func (icmp *Icmp) expireTransaction(tuple HashableIcmpTuple, trans *IcmpTransaction) {
+	trans.Notes = append(trans.Notes, OrphanedRequestMsg)
+	logp.Debug("icmp", OrphanedRequestMsg+" %s", &trans.Tuple)
+	icmp.publishTransaction(trans)
+}
+
+func (icmp *Icmp) publishTransaction(trans *IcmpTransaction) {
+	if icmp.results == nil {
+		return
+	}
+
+	logp.Debug("icmp", "Publishing transaction. %s", &trans.Tuple)
+
+	event := common.MapStr{}
+
+	// common fields - group "env"
+	event["client_ip"] = trans.Tuple.SrcIp
+	event["ip"] = trans.Tuple.DstIp
+
+	// common fields - group "event"
+	event["timestamp"] = common.Time(trans.Ts) // timestamp of the first packet
+	event["type"] = "icmp"                     // protocol name
+	event["count"] = 1                         // reserved for future sampling support
+	event["path"] = trans.Tuple.DstIp          // what is requested (dst ip)
+	if trans.HasError() {
+		event["status"] = common.ERROR_STATUS
+	} else {
+		event["status"] = common.OK_STATUS
+	}
+	if len(trans.Notes) > 0 {
+		event["notes"] = trans.Notes
+	}
+
+	// common fields - group "measurements"
+	responsetime, hasResponseTime := trans.ResponseTimeMillis()
+	if hasResponseTime {
+		event["responsetime"] = responsetime
+	}
+	switch icmp.direction(trans) {
+	case IcmpDirectionFromInside:
+		if trans.Request != nil {
+			event["bytes_out"] = trans.Request.Length
+		}
+		if trans.Response != nil {
+			event["bytes_in"] = trans.Response.Length
+		}
+	case IcmpDirectionFromOutside:
+		if trans.Request != nil {
+			event["bytes_in"] = trans.Request.Length
+		}
+		if trans.Response != nil {
+			event["bytes_out"] = trans.Response.Length
+		}
+	}
+
+	// common fields - group "raw"
+	// TODO: Add more info than only "type" & "code".
+	//       The IPv4/IPv6 payload could be interesting.
+	// if icmp.SendRequest {
+	// 	event["request"] = ""
+	// }
+	// if icmp.SendResponse {
+	// 	event["response"] = ""
+	// }
+
+	// event fields - group "icmp"
+	icmpEvent := common.MapStr{}
+	event["icmp"] = icmpEvent
+
+	icmpEvent["version"] = trans.Tuple.IcmpVersion
+
+	if trans.Request != nil {
+		icmpEvent["request_type"] = trans.Request.Type
+		icmpEvent["request_code"] = trans.Request.Code
+	}
+	if trans.Response != nil {
+		icmpEvent["response_type"] = trans.Response.Type
+		icmpEvent["response_code"] = trans.Response.Code
+	}
```

Comment:
See: #125
Plan is to fully restrict types for filtering support. This restricting might give us a chance to not use maps internally. 

Right now you can store any JSON marshable object in your event. Just top level event object needs to be MapStr.